pkg/relayer/proxy/relay_meter.go

```diff
@@ -270,105 +289,87 @@ func (rmtr *ProxyRelayMeter) AllowOverServicing() bool {
-	// max amount of stake the application can consume.
-	if !ok {
-		var app apptypes.Application
-		// Application stake is guaranteed to be up-to-date as long as the cache is
-		// invalidated at each new block.
-		app, err := rmtr.applicationQuerier.GetApplication(ctx, appAddress)
-		if err != nil {
-			return nil, err
-		}
+	// Build a new entry (same logic you already have)
+	appAddress := reqMeta.GetSessionHeader().GetApplicationAddress()
 
-		// In order to prevent over-servicing, the protocol must split the application's stake
-		// among all the suppliers that are serving it.
-		if len(app.ServiceConfigs) != 1 {
-			return nil, ErrRelayerProxyInvalidSession.Wrapf(
-				"application %q has %d service configs, expected 1",
-				appAddress,
-				len(app.ServiceConfigs),
-			)
-		}
+	// Application stake is guaranteed to be up-to-date as long as the cache is
+	// invalidated at each new block.
+	app, err := rmtr.applicationQuerier.GetApplication(ctx, appAddress)
+	if err != nil {
+		return nil, err
+	}
 
-		sharedParams, err := rmtr.sharedQuerier.GetParams(ctx)
-		if err != nil {
-			return nil, err
-		}
+	// In order to prevent over-servicing, the protocol must split the application's stake
+	// among all the suppliers that are serving it.
+	if len(app.ServiceConfigs) != 1 {
+		return nil, ErrRelayerProxyInvalidSession.Wrapf(
+			"application %q has %d service configs, expected 1",
+			appAddress,
+			len(app.ServiceConfigs),
+		)
+	}
 
-		sessionParams, err := rmtr.sessionQuerier.GetParams(ctx)
-		if err != nil {
-			return nil, err
-		}
+	sharedParams, err := rmtr.sharedQuerier.GetParams(ctx)
+	if err != nil {
+		return nil, err
+	}
 
-		// calculate the max amount of stake the application can consume in the current session.
-		supplierAppStake := getAppStakePortionPayableToSessionSupplier(
-			app.GetStake(),
-			sharedParams,
-			sessionParams.GetNumSuppliersPerSession(),
-		)
+	sessionParams, err := rmtr.sessionQuerier.GetParams(ctx)
+	if err != nil {
+		return nil, err
+	}
 
-		relayMeter = &sessionRelayMeter{
-			app:           app,
-			consumedCoin:  cosmostypes.NewInt64Coin(pocket.DenomuPOKT, 0),
-			maxCoin:       supplierAppStake,
-			sessionHeader: reqMeta.SessionHeader,
-		}
+	// calculate the max amount of stake the application can consume in the current session.
+	supplierAppStake := getAppStakePortionPayableToSessionSupplier(
+		app.GetStake(),
+		sharedParams,
+		sessionParams.GetNumSuppliersPerSession(),
+	)
+
+	relayMeter := &sessionRelayMeter{
+		app:           app,
+		consumedCoin:  cosmostypes.NewInt64Coin(pocket.DenomuPOKT, 0),
+		maxCoin:       supplierAppStake,
+		sessionHeader: reqMeta.SessionHeader,
+	}
 
-		rmtr.relayMeterMu.Lock()
-		rmtr.sessionToRelayMeterMap[sessionId] = relayMeter
-		rmtr.relayMeterMu.Unlock()
+	// Try to publish; if someone beat us, reuse theirs.
+	if existing, loaded := rmtr.sessionToRelayMeterMap.LoadOrStore(sessionId, relayMeter); loaded {
```

Comment:
`LoadOrStore` returns the newly created value if an entry does not exist.

Why not do the following?

```suggestion
	relayMeter = rmtr.sessionToRelayMeterMap.LoadOrStore(sessionId, relayMeter)
```